sqlglot/optimizer/qualify_columns.py

```diff
@@ -413,19 +488,46 @@ def _expand_stars(
             if not pivot_output_columns:
                 pivot_output_columns = [c.alias_or_name for c in pivot.expressions]
 
+    if dialect == "bigquery" and any(isinstance(col, exp.Dot) for col in scope.stars):
+        # Found struct expansion, annotate scope ahead of time
+        annotator.annotate_scope(scope)
+
     for expression in scope.expression.selects:
+        tables = []
+        structs = []
         if isinstance(expression, exp.Star):
             tables = list(scope.selected_sources)
             _add_except_columns(expression, tables, except_columns)
             _add_replace_columns(expression, tables, replace_columns)
-        elif expression.is_star and not isinstance(expression, exp.Dot):
-            tables = [expression.table]
-            _add_except_columns(expression.this, tables, except_columns)
-            _add_replace_columns(expression.this, tables, replace_columns)
+        elif expression.is_star:
+            if not isinstance(expression, exp.Dot):
+                tables = [expression.table]
+                _add_except_columns(expression.this, tables, except_columns)
+                _add_replace_columns(expression.this, tables, replace_columns)
+            elif dialect == "bigquery":
+                col = expression.find(exp.Column)
+                table_name = col.table if col else None
+                starting_struct = str(expression.this)
+                structs.append((col, starting_struct))
         else:
             new_selections.append(expression)
             continue
 
+        if structs:
+            struct_fields = _expand_struct_stars(scope, structs) or []
+
+            for struct_name in struct_fields:
+                root, *parts = struct_name.split(".")
+                alias_ = next_alias_name()
+                new_column: exp.Column | exp.Dot = exp.column(root, table=table_name)
+
+                if parts:
+                    new_column = exp.Dot.build(
+                        [new_column, *[exp.Identifier(this=part) for part in parts]]
```

Comment:
Need to be careful about quotes, this strips them and can break queries with unsafe identifiers used as struct fields.

Reply:
When gathering these `parts` I extract their text as is, so when stitching them back together here I think the backticks are preserved e.g.: 

```
sql = "WITH tbl1 AS (SELECT STRUCT(1 AS `from`) AS col) SELECT tbl1.col.* from tbl1";

result = qualify(parse_one(sql), ....).sql(...)

print(result)
```

```
WITH tbl1 AS (SELECT STRUCT(1 AS `from`) AS col) SELECT tbl1.col.`from` AS _field_0 FROM tbl1 AS tbl1
```

Is this fine or is the problem you're mentioning more nuanced?